### PR TITLE
KAFKA-4405: avoid calling pollNoWakeup unnecessarily

### PR DESCRIPTION
**Pull request by enothereska**

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: Nit: This should be called `maxPollRecords` as we don't use the `get` prefix for methods like this.

**Review by ijuma — COMMENTED** [review]
Thanks for the PR. I think we need to understand a little better what is the reason for the slowdown as this change may make some cases slower (if there is a higher latency between client and brokers for example).

Also, it may be worth testing after KAFKA-4469 is merged as that reduces some of the cost when `max.poll.records` is smaller than the fetch size.

**Comment by ijuma** [comment]
cc @hachikuji

**Comment by enothereska** [comment]
I'll test with KAFKA-4469 and report back. Thanks.

**Comment by enothereska** [comment]
@ijuma @hachikuji I tested with the new trunk containing KAFKA-4469 and I don't see the performance discrepancy anymore. So I'll close this PR. Thanks.

**Comment by ijuma** [comment]
Great news @enothereska, thanks for testing.

**Comment by enothereska** [comment]
Actually folks, it turns out I was testing with max.poll.records = 1000 by mistake. When testing with max.poll.records = 1, the issue remains and this PR is still needed. Sorry about the confusion.

**Comment by hachikuji** [comment]
Hey @enothereska, thanks for the patch. Just to clarify, with this fix, the performance discrepancy goes away totally?

**Comment by enothereska** [comment]
@hachikuji with this fix the perf when we use max.poll.records=1 goes up by 50%. However, the perf when we use max.poll.records=1000 is still higher. Example numbers: 10MB/s (mpr=1)->15MB/s(mpr=1 with this PR)->19MB/s(mpr=1000).

**Comment by enothereska** [comment]
Known failure, we are debating on whether to turn off that test while we sort it out. Unrelated to this PR.

**Comment by hachikuji** [comment]
@enothereska Thanks, the difference is puzzling... The description from the JIRA doesn't seem quite right because, in spite of the call to `sendFetches`,  no fetches will be sent for any partition which still has buffered records. Could the performance degradation be caused by the overhead of analyzing which fetches need to be sent? How many partitions are in your test case?

**Comment by ijuma** [comment]
@hachikuji I was thinking the same as you. For `max.poll.records=1`, the amount of work done by the user per `poll` could potentially be really low in comparison to the work to figure out which partitions need a `fetch`.

**Comment by ijuma** [comment]
Having said that, I am not sure if it ever makes sense to have such a low `max.poll.records` after KIP-62. It may be reasonable to say that if processing of each record is very cheap, you need to use a higher `max.poll.records` for best performance.

**Comment by hachikuji** [comment]
@ijuma Yeah, good point. If the processing is fast enough that `poll` itself is becoming a bottleneck, then it seems safe to increase `max.poll.records` to reduce the number of calls to `poll`. 

One thing I noticed in `Fetcher.fetchablePartitions` is that we're also doing an expensive call to `remove` from an `ArrayList`. We could probably be a bit more clever. In fact, just using a `HashSet` would probably be a nice improvement. Maybe @enothereska could try that and see if it makes a difference?

**Comment by enothereska** [comment]
If poll should be large, I wouldn't over-optimize this case.

**Comment by enothereska** [comment]
The topic has 1 partition.

**Comment by hachikuji** [comment]
Haha, that mostly rules out overhead from `Fetcher`. The only thing left is the `pollNoWakeup`.

**Comment by enothereska** [comment]
@hachikuji @ijuma I'll let you guys decide on this. Let me know what you want to do. If not needed, I can close the PR. Thanks.

**Comment by guozhangwang** [comment]
@hachikuji Could you let @eno know what are the next steps? Thanks.

**Comment by hachikuji** [comment]
@guozhangwang @enothereska From what I can tell so far, the main issue _is_ actually the call to `pollNoWakeup()` as I suggested above. One easy fix is to have `sendFetches` return a boolean indicating whether any fetches were sent. If nothing was sent, we can skip the call to `pollNoWakeup`. I tried this out and it does improve performance quite a bit.

**Comment by enothereska** [comment]
@hachikuji I tried that and it had little effect on the streams performance (+1MB/s or so, can't tell if it's significant).

**Comment by hachikuji** [comment]
@enothereska I loaded a single partition with 20,000,000 1K records. I used the consumer performance test. This is trunk:

|max.poll.records|throughput|
|-----------------|-----------|
|1|248.1233|
|16|513.6948|
|64|527.4310|
|65536|545.6583|

This is the results if we optimize the `pollNoWakeup()` call:

|max.poll.records|throughput|
|-----------------|-----------|
|1|394.8962|
|16|526.4992|
|64|529.6719|
|65536|537.3721|

And here is using this patch:

|max.poll.records|throughput|
|-----------------|-----------|
|1|410.0502|
|16|509.8636|
|64|528.4978|
|65536|527.4164|

There was a bit of variance in the individual tests, but it's tough to dispute the +150MB difference with max.poll.records=1 between trunk and the two alternatives on the table. My takeaways were the following:

1. Optimizing `pollNoWakeup` gets most of the benefit without changing prefetching behavior.
2. Increasing max.poll.records even a little gets you close to the maximum throughput possible.

I guess we should try to figure out why your results are coming out different.

**Comment by enothereska** [comment]
@hachikuji If 2. (increasing max.poll.records a little) gets most of the benefit, I propose we don't spend more time on this problem and close this PR.

**Comment by ijuma** [comment]
A PR to avoid the unnecessary pollNoWakeup seems worthwhile IMO.

**Comment by hachikuji** [comment]
@ijuma I agree, seems like an easy win. We can do it here or separately, whichever @enothereska prefers.

**Comment by guozhangwang** [comment]
I think we can use this PR to merge in the `pollNoWakeup` fix.

**Comment by enothereska** [comment]
Ok I can put that in. Thanks.

**Comment by enothereska** [comment]
@hachikuji can you check this is what you had in mind? Thanks.

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: This condition is reversed, we should call `pollNoWakeup` if a fetch has been sent.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: Why not return the number of fetches that were sent?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: While you're here, I suggest fixing `fetchablePartitions` so that it doesn't do two `remove` calls on an `ArrayList` (in the worst case, it has to shift all the elements in the underlying array twice). We could pass a predicate to `subscriptions.fetchablePartitions()` and make it more efficient.

**Review by ijuma — CHANGES_REQUESTED** [review]
Thanks, left a few suggestions.

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java
Comment (enothereska) [reply clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: doh, good catch!

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java
Comment (enothereska) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: Ok will do, but we probably won't use this info.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java
Comment (enothereska) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: Ok, I'll have a look.

Comment (enothereska) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: I took the easy way out with using a `HashSet`

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: It doesn't seem like we need this variable. Why not assign the result of `createFetchRequests` and call `size` on it? Less mutable variables that need to be tracked.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java]: Order must be maintained, so this changes the behaviour. You'd have to use a `LinkedHashSet`, which is a lot more expensive to create than an `ArrayList` and would probably defeat the point of the change. If we think the usage of a predicate is too awkward, we could replace the two `remove` calls with a single `removeAll`. That would probably be a little cheaper.

Also, we can probably pre-size the `ArrayList` in this method with `assignment.size()`. `isFetchable` should be true in most cases.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetcherTest.java
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetcherTest.java]: Seems like we should just say `assertEquals(0, fetcher.sendFetches)`. It would also be good to test the actual number in at least one of the non zero cases.

**Review by ijuma — COMMENTED** [review]
Thanks for the changes, left a few comments.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java]: Yeah, my fault for suggesting this earlier. I had forgotten about the significance of the order. Instead of a predicate, we could also pass in a set of the partitions to exclude. Switching to `removeAll` also sounds better.

**Comment by enothereska** [comment]
I can verify that the performance indeed increased as expected with this change.

**Comment by ijuma** [comment]
Thanks, LGTM. What do you think @hachikuji?

**Comment by hachikuji** [comment]
LGTM. Thanks for the patch!

**Pull request closed without merging**
